feature(cSDK): Unstructured file support example (#614)
fivetran/fivetran_connector_sdk: examples/quickstart/unstructured_data/README.md

Reviewer: We should get the review of tech writers on the format of this file

Author: I have pinged Chinmayi to review this PR.

# Unstructured Data Examples

## Overview

Learn how to upload unstructured data such as PDFs, images, and binaries to a destination using your Connector SDK code.

Unlike structured data, which has traditional tabular data (rows/columns) that fits into database tables, unstructured data includes files that don't fit neatly into rows and columns:
- PDFs (reports, invoices, contracts)
- Images (photos, screenshots, diagrams)
- Archives (ZIP, TAR)
- Any binary content

> Note: Unstructured file support is currently available for Snowflake, BigQuery, and Databricks destinations only. See [Unstructured File Replication](https://fivetran.com/docs/core-concepts/features/unstructured-file-replication) for more details.

---

## Getting started

> **New to unstructured data?** Start here:

Reviewer 2: ```suggestion
New to unstructured data? Start here:
```

-  [file_lifecycle](file_lifecycle/): Upload your first file and understand the upload, update, and delete lifecycle (5 minutes).
-  [stream_examples](stream_examples/): Learn different streaming approaches and when to use `expected_bytes`.

To run an example:

```bash
# 1. Pick an example
cd file_lifecycle

# 2. Run locally
fivetran debug
```
---

> ## How it works

Reviewer: we can add flow diagrams instead of these for the readme

Reviewer 2: +1

| 1       | report.pdf  | documents/report.pdf    |
```

#### Understanding file paths:

| What | Who Controls It | Example |
|------|-----------------|---------|
| `FileUpload.path` | You (connector code) | `"documents/report.pdf"` |
| `_fivetran_file_path` | Fivetran (stores the same value) | `"documents/report.pdf"` |

> Note: The path you provide in `FileUpload.path` is stored in `_fivetran_file_path`. They contain the same value.

---

## Examples

| I want to | Go to |
|-------------|-------|
| Upload my first file | [file_lifecycle](file_lifecycle/) |
| Understand the upload, update, and delete lifecycle | [file_lifecycle](file_lifecycle/) |
| Learn  the `_fivetran_file_path` behavior | [file_lifecycle](file_lifecycle/) |
| Learn different streaming approaches | [stream_examples](stream_examples/) |
| Understand when to use expected_bytes | [stream_examples](stream_examples/) |

---

## Common troubleshooting

### `_fivetran_file_path` column is missing
- ❌ Not passing `file=FileUpload(...)` parameter
- ✅ Check whether your `upsert()` or `update()` call  includes the `file` parameter

### "FileUpload object has no attribute 'read'"

> - ❌ Your stream object doesn't have a `read(size)` method

Reviewer: I do not think that we should use icons here 
Can you please check with tech writers regarding this?

Reviewer 2: +1

Author: I have removed the icons.

- ✅ Use `response.raw`, `BytesIO`, or file object

### File not appearing in destination
- ❌ Not passing `file=FileUpload(...)` to `Operations.upsert()`
- ✅ Check whether your `upsert()` call includes the `file` parameter

### "expected_bytes mismatch"
- ❌ File size changed during upload
- ✅ Verify `expected_bytes` matches the actual file size

---

> **Ready to get started?** Head to [file_lifecycle](file_lifecycle/) →

Reviewer 2: ```suggestion
Ready to get started? Head to [file_lifecycle](file_lifecycle/).
```

> # Unstructured Data Examples

Reviewer 2: There's an overall logical ordering issue with the sections. 
We can broadly have:
Overview
Getting Started
How it works
Examples

Author: Updated it.

> Learn how to sync files (PDFs, images, binaries, etc.) using the Fivetran Connector SDK.

Reviewer 2: ```suggestion
Learn how to upload unstructured data such as PDFs, images, and binaries to a destination using your Connector SDK code.
```

> > **Note:** Unstructured file support is currently available for Snowflake, BigQuery, and Databricks destinations only. See [Unstructured File Replication](https://fivetran.com/docs/core-concepts/features/unstructured-file-replication) for more details.

Reviewer 2: ```suggestion
> Note: Unstructured file support is currently available for Snowflake, BigQuery, and Databricks destinations only. See [Unstructured File Replication](https://fivetran.com/docs/core-concepts/features/unstructured-file-replication) for more details.
```

---

## Quick start

**New to unstructured data?** Start here:

1. **[file_lifecycle](file_lifecycle/)** - Upload your first file + understand upload/update/delete lifecycle (5 minutes)

> 2. **[stream_examples](stream_examples/)** - Learn different streaming approaches and when to use expected_bytes

Reviewer 2: ```suggestion
-  [file_lifecycle](file_lifecycle/): Upload your first file and understand the upload, update, and delete lifecycle (5 minutes).
-  [stream_examples](stream_examples/): Learn different streaming approaches and when to use `expected_bytes`.
```

---

> ## What is unstructured data?

Reviewer 2: We can move this section to the overview.

Author: Updated it.

> **Unstructured data** = Files that don't fit neatly into rows and columns:

Reviewer 2: ```suggestion
Unlike structured data, which has traditional tabular data (rows/columns) that fits into database tables, unstructured data includes files that don't fit neatly into rows and columns:
```

- PDFs (reports, invoices, contracts)
- Images (photos, screenshots, diagrams)
- Archives (ZIP, TAR)
- Any binary content

> **Structured data** = Traditional tabular data (rows/columns) that fits into database tables.

Reviewer 2: ```suggestion

```

---

## How it works

```
Your Connector
  ↓
  Downloads file from source API
  ↓
  Creates FileUpload(path="file.pdf", stream=...)
  ↓
  Calls Operations.upsert(table, data, file=...)
  ↓
Fivetran Platform
  ↓
  Loads the file to destination storage
  ↓
  Writes row to destination with _fivetran_file_path
  ↓
Destination (Snowflake/BigQuery/Databricks)
  ↓
  Table contains metadata columns + _fivetran_file_path
  ↓
  You can query/download files using the path
```

---

## The `_fivetran_file_path` column

When you upload a file, Fivetran **automatically adds** a `_fivetran_file_path` column to your table to track the file location.

> **You don't define this column** - it's added by the platform automatically.

Reviewer 2: ```suggestion
When you upload a file, Fivetran automatically adds a `_fivetran_file_path` column to your table to track the file location. You don't define this column; the platform adds it automatically.
```

### Example:

> **Your schema:**

Reviewer 2: ```suggestion
Your schema:

```

```python
def schema(configuration):
    return [{
        "table": "documents",
        "primary_key": ["doc_id"],
        "columns": {
            "doc_id": "STRING",
            "doc_name": "STRING",
        },
    }]
```

> **After sync, query your destination:**

Reviewer 2: ```suggestion
After sync, query your destination:

```

```sql
SELECT doc_id, doc_name, _fivetran_file_path FROM documents;
```

> **Result:**

Reviewer 2: ```suggestion
Result:

```

```
| doc_id  | doc_name    | _fivetran_file_path     |
|---------|-------------|-------------------------|
| 1       | report.pdf  | documents/report.pdf    |
```

### Understanding file paths:

| What | Who Controls It | Example |
|------|-----------------|---------|
| `FileUpload.path` | **You** (connector code) | `"documents/report.pdf"` |

> | `_fivetran_file_path` | **Fivetran** (stores the same value) | `"documents/report.pdf"` |

Reviewer 2: ```suggestion
| `FileUpload.path` | You (connector code) | `"documents/report.pdf"` |
| `_fivetran_file_path` | Fivetran (stores the same value) | `"documents/report.pdf"` |
```

> **Key insight:** The path you provide in `FileUpload.path` is stored in `_fivetran_file_path`. They contain the same value.

Reviewer 2: ```suggestion
> Note: The path you provide in `FileUpload.path` is stored in `_fivetran_file_path`. They contain the same value.
```

---

## Choose your example

> | I want to... | Go to |

Reviewer 2: ```suggestion
| I want to | Go to |
```

|-------------|-------|
| Upload my first file | [file_lifecycle](file_lifecycle/) |

> | Understand upload/update/delete lifecycle | [file_lifecycle](file_lifecycle/) |

Reviewer 2: ```suggestion
| Understand the upload, update, and delete lifecycle | [file_lifecycle](file_lifecycle/) |
```

> | Learn `_fivetran_file_path` behavior | [file_lifecycle](file_lifecycle/) |

Reviewer 2: ```suggestion
| Learn  the `_fivetran_file_path` behavior | [file_lifecycle](file_lifecycle/) |
```

| Learn different streaming approaches | [stream_examples](stream_examples/) |
| Understand when to use expected_bytes | [stream_examples](stream_examples/) |

---

## Key concepts covered

### [file_lifecycle/](file_lifecycle/)
- ✅ How to upload a file
- ✅ How to update a file (upsert vs update)
- ✅ How to delete a file
- ✅ `_fivetran_file_path` lifecycle
- ✅ State management with checkpoint

### [stream_examples/](stream_examples/)
- ✅ HTTP response streams (requests.raw)
- ✅ In-memory bytes (BytesIO)
- ✅ File handles (local files)
- ✅ With expected_bytes (size validation)

> - ✅ Without expected_bytes (unknown size)

Reviewer 2: ```suggestion

```

---

## Common troubleshooting

### `_fivetran_file_path` column is missing
- ❌ Not passing `file=FileUpload(...)` parameter

> - ✅ Check your upsert() or update() call includes `file` parameter

Reviewer 2: ```suggestion
- ✅ Check whether your `upsert()` or `update()` call  includes the `file` parameter
```

### "FileUpload object has no attribute 'read'"
- ❌ Your stream object doesn't have a `read(size)` method
- ✅ Use `response.raw`, `BytesIO`, or file object

### File not appearing in destination
- ❌ Not passing `file=FileUpload(...)` to `Operations.upsert()`

> - ✅ Check your upsert() call includes `file` parameter

Reviewer 2: ```suggestion
- ✅ Check whether your `upsert()` call includes the `file` parameter
```

### "expected_bytes mismatch"
- ❌ File size changed during upload

> - ✅ Verify expected_bytes matches actual file size

Reviewer 2: ```suggestion
- ✅ Verify `expected_bytes` matches the actual file size
```

---

## Key requirements

### 1. FileUpload with Stream
```python
FileUpload(
    path="documents/filename.pdf",  # Path within table namespace
    stream=file_stream               # Stream with read(size) method
)
```

### 2. Operations.upsert or update with file Parameter
```python
Operations.upsert(
    table="documents",
    data={"doc_id": "1", "doc_name": "report.pdf"},
    file=FileUpload(...)  # ← Pass file here
)
```

---

## Getting started

```bash
# 1. Pick an example
cd file_lifecycle

# 2. Run locally
fivetran debug
```

---

> ## Pro tips

Reviewer 2: ```suggestion
## Tips
```

1. **Start with file_lifecycle/** to understand fundamentals
2. **Use stream_examples/** to learn different streaming approaches
3. **Test locally first** with `fivetran debug` before deploying

> 4. **Check destination** after sync to verify `_fivetran_file_path` is populated

Reviewer 2: ```suggestion
1. Start with `file_lifecycle/` to understand fundamentals
2. Use `stream_examples/` to learn different streaming approaches
3. Test locally first with `fivetran debug` before deploying
4. Check destination after sync to verify `_fivetran_file_path` is populated
```

## How it works

> The file upload process follows these steps:

Reviewer 2: Uploading a file involves the following steps:

Author: Removed the points and just kept the flow diagram.

> 1. **Your Connector** fetches the file from the source (API, Cloud storage)

Reviewer 2: 1. Your connector fetches the file from the source (API, cloud storage).                               
  2. Create a `FileUpload` object with the path for the file in your destination and a stream to read  
  from (`FileUpload(path="file.pdf", stream=...)`).                                                      
  3. Call `upsert()` with the file parameter to upload both the file and metadata (`op.upsert(table,     
  data, file=...)`).                                                                                     
  4. The Fivetran platform receives the file and metadata from your connector.                         
  5. The Fivetran platform stores the file in the destination's file storage (Snowflake stages, BigQuery
  Cloud Storage, or Databricks volumes).
  6. The Fivetran platform writes a metadata row to the destination table, including the automatically
  added `_fivetran_file_path` column with the file path.
  
To download the file from your destination, query the table to retrieve the `_fivetran_file_path` value, then use that value to download the file.

Reviewer 2: I would prefer it to be a diagram because there are different actors involved. It's not coming out clearly in the written steps. For example, do steps 1 and 2 happen concurrently or one after the other?
